field/TinyLang/Field/UParser.hs

```diff
@@ -0,0 +1,348 @@
+identifierCharLabel = "identifier character [a-z0-9_']"
+
+identifierChar :: (MonadParsec e s m, Token s ~ Char) => m (Token s)
+identifierChar = satisfy isIdentifierChar <?> identifierCharLabel
+
+keyword :: String -> Parser ()
+keyword kwd = lexeme (string kwd *> notFollowedBy identifierChar)
+
+keywords :: Set String
+keywords =
+    fromList
+    [ "T", "F"
+    , "not", "and", "or", "xor"
+    , "neq0", "neg", "inv"
+    , "let"
+    , "if", "then", "else"
+    , "for", "do", "end"
+    , "unpack"
+    ]
+
+isKeyword :: String -> Bool
+isKeyword = (`member` keywords)
+
+pIdentifier :: Parser Identifier
+pIdentifier =
+    lexeme $ do
+        prefix     <- option "" (string "?" <|> string "#")
+        identifier <- (:) <$> lowerChar
+                          <*> takeWhileP (Just identifierCharLabel)
+                              isIdentifierChar
+        pure $ prefix ++ identifier
+
+pBoolLiteral :: Parser Bool
+pBoolLiteral =
+    lexeme $ charToBool <$> (satisfy isTF <?> "T or F")
+    where
+        isTF x = x == 'T' || x == 'F'
+        --
+        charToBool 'T' = True
+        charToBool 'F' = False
+        charToBool _   = error "impossible"
+
+pIntLiteral :: Parser Integer
+pIntLiteral =
+    lexeme $ fmap read (some digitChar)
+
+pVecLiteral :: Parser [Bool]
+pVecLiteral =
+    lexeme $
+        between
+            (symbol "{")
+            (symbol "}")
+            (pBoolLiteral `sepBy` (symbol ","))
+
+pConst :: Parser Const
+pConst =
+    choice
+        [ CBool <$> pBoolLiteral
+        , CInt  <$> pIntLiteral
+        , CVec  <$> pVecLiteral
+        ]
+
+-- variable is an identifier that is not a keyword
+pVar :: Parser Var
+pVar = do
+    ident <- pIdentifier
+    when (isKeyword ident)
+         (fail ("keyword " ++ show ident ++ " cannot be an identifier"))
+    pure $ Var ident
+
+parens :: Parser a -> Parser a
+parens = between (symbol "(") (symbol ")")
+
+unary :: Parser a -> UnOp -> Comb.Operator Parser Expr
+unary pName op = Comb.Prefix (EAppUnOp op <$ pName)
+
+binary :: Parser a -> BinOp -> Comb.Operator Parser Expr
+binary pName op = Comb.InfixL (EAppBinOp op <$ pName)
+
+operatorTable :: [[Comb.Operator Parser Expr]]
+operatorTable =
+    [ [ unary  (keyword "not")    $ Not
+      ]
+    , [ unary  (keyword "neg")    $ Neg
+      , unary  (keyword "inv")    $ Inv
+      ]
+    , [ binary (symbol  "*")      $ Mul
+      , binary (symbol  "/")      $ Div
+      ]
+    , [ binary (symbol  "+")      $ Add
+      , binary (symbol  "-")      $ Sub
+      ]
+    , [ binary (symbol  "==")     $ FEq
+      , binary (symbol  "<=")     $ FLe
+      , binary (symbol  "<")      $ FLt
+      , binary (symbol  ">=")     $ FGe
+      , binary (symbol  ">")      $ FGt
+      , unary  (keyword "neq0")   $ Neq0
+      ]
+    , [ binary (keyword "xor")    $ Xor
+      , binary (keyword "and")    $ And
+      , binary (keyword "or")     $ Or
+      ]
+    , [ unary  (keyword "unpack") $ Unp
+      ]
+    ]
```

Comment:
@effectfully, what is the operator precedence for this language?

The original spec mentions the following precedences:
- `not > xor > and > or`, and
- `{neg,inv} > {*,/} > {+,-}`

but it does not mention operator precedence for the following
- `neq0`
- `==, <=, <, >=, >`
- `unpack`
- `expr [ expr ]`

So far I assumed that:
- `not` > everything
- arithmetic > `neq0`,
- arithmetic > boolean, and 
- everything  > unpack.

Reply:
All prefix unary operators (`not`, `neq0`, `neg`, `inv`, `unpack`) should have the same precedence that is greater than the precedences of all the binary operators. And both `unOp1 (unOp2 x)` and `unOp1 (x binOp2 y)` always require parentheses. `unOp1 x binOp unOp2 y` never requires parentheses.

`expr [ expr ]` is a postfix operator. It should have precedence that is less than the one of unary operators, but still greater than the precedences of all the binary operators. So that `unOp x [ expr ]` never requires parentheses, but `(x binOp y) [ expr ]` always requires them.

Binary operators are handled as per usual like you've assumed.

Reply:
What is the difference between:
- `unOp1 (x binOp2 y)`, and
- `unOp1 x binOp2 y`
?

Reply:
Sorry for being unclear.

- `unOp1 (x binOp2 y)` parses as `unOp1 (x binOp2 y)`
- `unOp1 x binOp2 y` parses as `(unOp1 x) binOp2 y`

Reply:
Err,

> Implements the order specified by @effectfully, however it does permit
the following syntax `a + b [c]` where `[c]` is binding as a postfix operator to `b`.

by my "`(x binOp y) [ expr ]` always requires parentheses" I mean that if the user wants an expression to parse like that, then parentheses are required. If the user wants to parse an expression like `x binOp (y [ expr ])`, then parentheses are not required and both of

- `x binOp y [ expr ]`
- `x binOp (y [ expr ])`

are fine. This is what I wanted to say by

> expr [ expr ] is a postfix operator. It should have precedence that is <...> greater than the precedences of all the binary operators. 

And this same logic applies to other precedence-related rules that I mentioned.

So what you implement is exactly what is desired, there is no "however" there.